### PR TITLE
dataflow: Make const-prop a true must-analysis

### PR DESCRIPTION
**Pull request by IagoAbal**
And only use reachable nodes when computing the input mapping of a node.

test plan:
make test

PR checklist:
- [x] Documentation is up-to-date
- [x] Changelog is up-to-date
- [x] Change has no security implications (otherwise, ping security team)

### CODE REVIEW

File: semgrep-core/src/analyzing/Dataflow_constness.ml [hunk 1]
```diff
@@ -280,22 +280,27 @@ and eval_concat env args =
 (* Transfer *)
 (*****************************************************************************)
 
-(* FIXME: This takes "Bottom" as the default constness of a variable.
```
Comment (aryx) [inline semgrep-core/src/analyzing/Dataflow_constness.ml:283]: I would have removed the FIXME but kept most of the comment, especially the example.

File: semgrep-core/src/analyzing/Dataflow_constness.ml [hunk 2]
```diff
@@ -280,22 +280,27 @@ and eval_concat env args =
...
+  let node = flow.graph#nodes#assoc ni in
+  match node.F.n with
+  | Enter -> enter_env
+  | _else -> (
+      let pred_envs =
+        CFG.predecessors flow ni
+        |> Common.map (fun (pi, _) -> mapping.(pi).D.out_env)
+      in
+      match pred_envs with
+      | [] -> VarMap.empty
```
Comment (aryx) [inline semgrep-core/src/analyzing/Dataflow_constness.ml:330]: why can't we simplify and just call fold_left with VarMap.empty as the accumulator?
Why we need the 3 cases? union_env will not work as intended then?

File: semgrep-core/src/core/il/CFG.ml [hunk 3]
```diff
@@ -15,12 +15,34 @@
...
+        (graph#successors nodei)#fold
+          (fun s (ni, _) -> NodeiSet.add ni s)
+          NodeiSet.empty
+      in
+      NodeiSet.fold aux succs seen
+  in
+  { graph; entry; reachable = aux entry NodeiSet.empty }
+
+(* Predecessors of a node (that can be reached from the entry node). *)
+let predecessors cfg nodei : (nodei * 'node) list =
```
Comment (aryx) [inline semgrep-core/src/core/il/CFG.ml:46]: I would then add a successor function too, just for symmetry.

File: semgrep-core/tests/python/constprop_dataflow.py [hunk 4]
```diff
@@ -1,28 +1,42 @@
...
+        # ERROR:
+        foo(a)
     # ERROR:
-    v = a
+    foo(a)
 
+def test3(c):
+    if c:
+        x = "hi"
+    # Must analysis, and `x`` is only coonstant in one path!
```
Comment (aryx) [inline semgrep-core/tests/python/constprop_dataflow.py]: typo, coonstant

**Review by aryx — APPROVED** [review]

File: semgrep-core/src/analyzing/Dataflow_constness.ml (hunk 1 above)
Comment (IagoAbal) [reply semgrep-core/src/analyzing/Dataflow_constness.ml:283]: Good point, will do.

**Pull request merged**